Create export_tsvs.py

bw2 opened the pull request:
This hail script takes the hail table generated by Mike's script and outputs one .tsv for each sample which contains all the variants that will be used from that sample.

bw2 committed:
Create export_tsvs.py

This hail script takes the hail table generated by Mike's script and outputs one .tsv for each sample which contains all the variants that will be used from that sample.

bw2 committed:
added checkpoint

bw2 committed:
fixed reference to S column

bw2 committed:
Switched to parallel tsv export. Also, better logging and var names.

bw2 committed:
Update export_tsvs.py

mike-w-wilson commented on export_tsvs.py:
```diff
@@ -0,0 +1,63 @@
+
+# combine ht.samples_w_het_var, ht.samples_w_hom_var, ht.samples_w_hemi_var into a single ht.samples array
+ht = ht.annotate(samples=ht.samples_w_het_var.extend(ht.samples_w_hom_var.extend(ht.samples_w_hemi_var)))
+ht = ht.drop(ht.samples_w_het_var, ht.samples_w_hom_var, ht.samples_w_hemi_var)
```
small thing but this could be a bit more concise with 
`ht = ht.select(samples=ht.samples_w_het_var.extend(ht.samples_w_hom_var.extend(ht.samples_w_hemi_var))`

bw2 replied:
oh yeah, nice

bw2 committed:
fix tsv export

mike-w-wilson commented on export_tsvs.py:
```diff
@@ -0,0 +1,64 @@
+
+# parse sample ids
+sample_ids = []
+with hl.hadoop_open(args.sample_ids_file_path) if args.sample_ids_file_path.startswith("gs://" ) else open(args.sample_ids_file_path, "rt") as f:
```
Another small thing but you use this `args.sample_ids_file_path` quite often, maybe declare it after parsing? something like?

`sample_ids_path = args.sample_ids_file_path`

mike-w-wilson commented on export_tsvs.py:
```diff
@@ -0,0 +1,64 @@
+p.add_argument("sample_ids_file_path", help="A text file containing one sample id per line")
+args = p.parse_args()
+
+# parse sample ids
```
Could you expand on this comment and the choice of i == 20? maybe that could be an import param?

bw2 replied:
This is just a logging detail, so doesn't seem worth a separate parameter. The linux `head` command prints 10 items by default if you run it without -n, and here I just picked 20. I'll change it to 10.

mike-w-wilson commented on export_tsvs.py:
```diff
@@ -0,0 +1,64 @@
+ht = ht.annotate(samples=ht.samples_w_het_var.extend(ht.samples_w_hom_var.extend(ht.samples_w_hemi_var)))
+ht = ht.drop(ht.samples_w_het_var, ht.samples_w_hom_var, ht.samples_w_hemi_var)
+
+# explode
```
I'm not sure these small comments are necessary.

mike-w-wilson reviewed (commented):
This looks good to me overall! My comments are mostly style things. One other comment - could you put this in a main function? It would help with readability.

bw2 commented:
> This looks good to me overall! My comments are mostly style things. One other comment - could you put this in a main function? It would help with readability.

Yes definitely.

bw2 committed:
address PR comments, split into functions

Also, added naive_coalesce to control number of output shards per sample vs. execution time. 
With --output-partitions=1000 , it currently takes 1 to 1.5 minutes per sample, so I'm planning to spin up many hail clusters in parallel, where each one generates something like ~500 .tsvs

gtiao commented on export_tsvs.py:
```diff
@@ -1,64 +1,138 @@
+	logging.info("Input schema:")
+	ht.describe()
+
+	ht = ht.annotate(
```
This is a very minor change to the structure in the input HT -- can we not just store this information in the desired format in the readviz prep script?

bw2 replied:
Mike did originally, but @lfrancioli suggested it was more space-efficient to store 3 arrays since it's not storing the extra het_or_hom_or_hemi field (which at the time was a string).

gtiao commented on export_tsvs.py:
```diff
@@ -1,64 +1,138 @@
+			per_sample_ht = per_sample_ht.naive_coalesce(n_partitions_per_sample)
+			per_sample_ht.export(tsv_output_path, parallel="separate_header")
+		else:
+			tsv_output_path = os.path.join(output_bucket_path, f"{s}.tsv.bgz")
```
I think if the input ht has multiple partitions you will need to still explicitly specify `per_sample_ht = per_sample_ht.naive_coalesce(1)` -- otherwise after filtering to the sample, Hail will maintain the original number of partitions from the input ht (unless things have changed recently)

bw2 replied:
In this `else` it writes out a single .tsv file, so I don't see where the naive_coalesce would help.

bw2 committed:
remove AC0 variants and also drop key and transmute into desired columns after naive_coalesce

bw2 committed:
remove duplicate describe()

bw2 committed:
bash script for running export_tsvs.py on many small dataproc clusters

export_tsvs.py takes 15 seconds per sample id (once the input table is reshuffled and keyed by sample). 
This running time doesn't improve with cluster size because the ht has ~50,000 partitions so on average all data for one sample sits on a single partition. This means all samples can be processed in ~10 hours on 20 small dataproc clusters which have 2 worker nodes each. This script creates these 20 clusters, submits jobs, and later deletes the clusters.

It may be possible to improve per-sample runtime by first repartitioning the table to ~500k partitions, using clusters with more worker nodes, and exporting sharded tsvs instead of a single .tsv.bgz. But these .tsvs would be more difficult to handle in the downstream step.

bw2 committed:
rekey by sample and switch back to exporting a single .tsv.bgz per sample since it now takes 15sec

Also, add `--start-with-sample-i` and `--n-samples-to-process` args to optionally  process only a subset of the sample ids. 
This allows a sample ids file to be divided up between multiple runs of export_tsvs.py without having to create subset files.
